### DISCUSSION TITLE
How to properly access properties & getters using LintRuleNodeRegistry (context.registry.addXXX) ?

### DISCUSSION THREAD

**Discussion by evaisse** [invertase/dart_custom_lint#213] [Q&A] [answered]
Hello, firstly, thank for this very usefull package, 

I try to detect dangerous usage of iterator methods which are not null safety, list `list.first`, `list.last`, `list[2]`.

Using the context.registry LintRuleNodeRegistry, what's the right listener to detect call of getters on an given object. 

For example, regarding the following code : 

```dart
final myList = <String>[];

void myTest() {
  final a = myList.first; //  that's what I try to detect
  final b = myList[42]; //  that's what I try to detect
  a.replaceFirst("a", "b"); //  this one is correctly detected by `addMethodInvocation`
}
```

For example I try to detect the `myList.first` node, and ensure that `myList` is instance of `List`. Using `context.registry.addIdentifier` I did found the right node, but I'm not sure I can get the correlation between the target and the method/propertyName.

`node.childEntities` give me an approximate answer like : `[myList, ., first]`

Is there any better way to get an invocation of an object getter or simply access to object property ? 

thx

**Comment by rrousselGit** [comment, marked as answer]
You can look for `Identifier` & subclasses for `myList.first`.
I don't remember what's for `[x]` operators, but that's likely going to be an `Expression` subclass.

One thing you can do to help debug is use a low-level `addNode` or `addExpression` or another general interface, and log `print('$node ${node.runtimeType}')`.
Then you'd know what object type you're looking for.

**Reply by evaisse** [reply]
Great, thank you, I found out that `addPrefixedIdentifier` did the job for getting the getter, and `addMethodInvocation` for method call.
Do you know if there is some kind of cookbook or reference ? I could add some hints, since I'm looking for pretty much any kind of expression.

**Reply by rrousselGit** [reply]
I'm not aware of one. The analyer API isn't that well documented, and constantly evolves as Dart changes

**Reply by evaisse** [reply]
Ok would you allow me to add some samples to the exemple project or to extends the doc as a quick cookbook ? 

Do you known if there is a way in the current state of the analyzer to detect function that throws ? I suppose any external lib will be ignored by the analyzer, but I can manage to detect throws expression in current analyzed scope.

**Reply by rrousselGit** [reply]
Depends on the sample/example/doc. I'd like to avoid having `custom_lint` serve as a documentation for `analyzer`

If analyzer needs doc and you want to make PRs, it'd probably make more sense to contribute to the analyzer docs directly.
